### ISSUE TITLE
Edge masking in GNNExplainer implementation

### ISSUE THREAD

**Issue by Gori-LV** [pyg-team/pytorch_geometric#4100]
### 🐛 Describe the bug

I found that edge masking seems to be missing in [implementation of GNNExplainer](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/models/gnn_explainer.py) in PyG.

```python
for epoch in range(1, self.epochs + 1):
    optimizer.zero_grad()
    h = x * self.node_feat_mask.sigmoid()
    out = self.model(x=h, edge_index=edge_index, **kwargs)
    if self.return_type == 'regression':
        loss = self.__loss__(mapping, out, prediction)
    else:
        log_logits = self.__to_log_prob__(out)
        loss = self.__loss__(mapping, log_logits, pred_label)
    loss.backward()
    optimizer.step()
```
This is the loop of explaining a node, line 286-298 in  the [source code](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/models/gnn_explainer.py). If my understanding is correct, only node features are masked by `h = x * self.node_feat_mask.sigmoid()`, while edge_index, i.e., the adjacency matrix is not. 

In the [official implementation](https://github.com/RexYing/gnn-model-explainer/blob/master/explainer/explain.py) by the author of GNNExplainer, it is done by the ` _masked_adj()` function. Due to the difference in data structure, I modify the code as below to do edge masking:
```python
for epoch in range(1, self.epochs + 1):
    optimizer.zero_grad()
    h = x * self.node_feat_mask.sigmoid()
    ### fix
    edge_weight = ori_edge_weight * self.edge_mask.sigmoid()
    out = self.model(x=h, edge_index=edge_index, edge_weight = edge_weight)
    ###
    # out = self.model(x=h, edge_index=edge_index, **kwargs)
    if self.return_type == 'regression':
        loss = self.__loss__(mapping, out, prediction)
    else:
        log_logits = self.__to_log_prob__(out)
        loss = self.__loss__(mapping, log_logits, pred_label)
    loss.backward()
    optimizer.step()
```
and add 
```python
if 'edge_weight' in kwargs:
    # print(kwargs['edge_weight'].shape)
    ori_edge_weight = kwargs['edge_weight']
    # print('edge_weight is not None')
else:
    ori_edge_weight = torch.ones((edge_index.size(1), ), dtype=edge_index.dtype,
                             device=edge_index.device)
    # print('edge_weight was None originally')
```
after `self.__subgraph__()` is called in line 261. 

Please let me know if this is incorrect and that I have misunderstood something :) btw, I did not try modifying the `explain_graph()` function but I suppose it should be a similar case.

Thanks a lot.

### Environment

* PyG version: 2.0.2
* PyTorch version: 1.8.0
* OS: MacOS
* Python version: 3.8
* How you installed PyTorch and PyG (`conda`, `pip`, source): `pip`

**Comment by rusty1s** [comment]
Since not all GNN operators utilize `edge_weight`, we apply the (soft) masking directly inside our `MessagePassing` interface, see [here](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/models/gnn_explainer.py#L86-L103) and [here](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/conv/message_passing.py#L326-L337). This makes `GNNExplainer` applicable to all GNN operators available in PyG.

**Comment by Gori-LV** [comment]
> Since not all GNN operators utilize `edge_weight`, we apply the (soft) masking directly inside our `MessagePassing` interface, see [here](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/models/gnn_explainer.py#L86-L103) and [here](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/conv/message_passing.py#L326-L337). This makes `GNNExplainer` applicable to all GNN operators available in PyG.

That's so cool! I did miss these parts. I'm actually trying to use GNNExplainer in heterogeneous GNN (e.g., [HTG](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/hetero/hgt_dblp.py)). Due to difference in data structure, I figure I need to modify [here](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/models/gnn_explainer.py#L76-L90) to initiate several edge masks (stored in a `dict`, perhaps) _w.r.t._ each edge type in the `data.edge_store`. Then change [here](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/models/gnn_explainer.py#L148-L151) to call `set_masks()` for each single edge mask. At the meantime, add node masks _w.r.t._ the node types. Would this setting allow GNNExplainer to be applicable to HGT?

**Comment by rusty1s** [comment]
Yes, that is correct. You need to take care of setting the correct mask for each individual type. Otherwise, everything else should stay the same.

**Comment by Gori-LV** [comment]
thanks a lot!

**Comment by Everyday-seu** [comment]
> thanks a lot!

Hi！
I‘m studying how to use GNNExplainer in HGT. 
Do you fix this problem now?
Hope for your reply!Thanks!
